A7-backward-compatibility-management.md

```diff
@@ -0,0 +1,94 @@
+
+### Related Proposals: 
+N/A
+
+## Proposal
+The major considerations of the design are:
+* Backward compatibility issues should be managed by gRPC server instead of user
+* Servers that do not care about any backward compatibility issue would not need to do anything special
+* Servers that care about a particular backward compatibility issue can invoke an API of gRPC library to configure the server so that the server will apply workaround to that issue based on client's version.
+
+### Client version identification
+gRPC server should identify the version of a client by analysing the `user-agent` header field on the server side. Current format of `user-agent` is well defined by the doc [PROTOCOL-HTTP2](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md). With this definition we can extract gRPC language and version from client's `user-agent` string and identify user's compatibility issues.
+
+### Server backward compatibility management
+gRPC server will maintain a list of server workarounds for backward compatibility issues. These workarounds are disabled by default for performance consideration. If a user cares about a backward compatibility issue, they should be able to enable the workaround. In this case gRPC server will identify the version of each client and determine whether the workaround should be applied.
+
+#### Workaround list
+The workaround list on gRPC server is a map from an integer `workaround id` to a `version comparing function`:
+
+| Workaround id | Version comparing function |
+| ------------- | -------------------------- |
+| 1             | `bool higher_than_v_1_2(uint32_t id, grpc_slice user_agent)` |
+| 2             | `bool not_objc_or_higher_than_v_1_3(uint32_t id, grpc_slice user_agent)` |
+| ...           | ... |
+
+* `Version comparing function` should be a function which takes a client `user-agent` string as parameter and returns a bool value indicating whether a workaround should be applied
+* The workaround list should be well documented in gRPC repository. The doc should include the items of the list, as well as detailed description of each workaround and its corresponding compatibility issue, so that users can identify those workarounds they want to enable
+* The list should be compiled into gRPC server and shipped to users in releases
+* An item is add to the list when a backward compatibility issue is raised and a workaround is implemented on the server
+* Updated list will be pushed to users in the next release (or next import for Google3)
+* An item should be removed from the list after 6 months (roughly 4 minor version releases)
+
+The list should be implemented and maintained by the three core languages for interoperability. Wrapping languages should use surface APIs (below) to access the table.
+
+#### APIs
+```
+GRPCAPI void enable_workaround(uint32_t id);
```

Comment:
Skip discussing the core API's here: instead talk about what abstract API's an implementation must export.

Example:

When building a server, an implementation should support enabling/disabling workarounds by ID. Those ID's should be listed in an enumeration.